File: cpp/day-03.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <ranges>
#include <unordered_map>
#include <sstream>
#include <utility>
#include <regex>
#include <stdint.h>

auto part_1(std::string &inst) -> uint64_t
{
    uint64_t ans = 0;
    std::regex word_regex(R"(mul\(\d+,\d+\))");
    auto words_begin = std::sregex_iterator(inst.begin(), inst.end(), word_regex);
    auto words_end = std::sregex_iterator();

    for (auto it = words_begin; it != words_end; it++)
    {
        auto match = it->str();
        long long num1 = 0, num2 = 0;
        int i = 4;
        while (match[i] != ',')
        {
            num1 = num1 * 10 + match[i] - '0';
            i++;
        }
        i++;
        while (match[i] != ')')
        {
            num2 = num2 * 10 + match[i] - '0';
            i++;
        }
        ans += num1 * num2;
    }
    return ans;
}

auto part_2(std::string &inst) -> uint64_t
{
    uint64_t ans = 0;
    std::regex word_regex(R"(mul\(\d+,\d+\)|do\(\)|don't\(\))");
    auto words_begin = std::sregex_iterator(inst.begin(), inst.end(), word_regex);
    auto words_end = std::sregex_iterator();

    bool enabled = true;
    for (auto it = words_begin; it != words_end; it++)
    {
        auto match = it->str();

        if (match == "do()")
        {
            enabled = true;
            continue;
        }
        if (match == "don't()")
        {
            enabled = false;
            continue;
        }
        if (!enabled)
            continue;
        
        int i = 4;
        long long num1 = 0, num2 = 0;
        while (match[i] != ',')
        {
            num1 = num1 * 10 + match[i] - '0';
            i++;
        }
        i++;
        while (match[i] != ')')
        {
            num2 = num2 * 10 + match[i] - '0';
            i++;
        }
        ans += num1 * num2;
    }
    return ans;
}

auto main(int argc, char **argv) -> int
{
    std::string inst;
    std::string line;
    while (std::getline(std::cin, line))
    {
        inst += line;
    }

    std::cout << "Part 1: " << part_1(inst) << std::endl;
    std::cout << "Part 2: " << part_2(inst) << std::endl;
    return 0;
}

File: rust/src/bin/day-03.rs
use std::io;
use regex::Regex;

fn main() {
    let data = io::stdin()
        .lines()
        .filter_map(Result::ok)
        .map(|x| x + " ")
        .collect::<String>();

        println!("Part 1: {}", part_one(&data));
        println!("Part 2: {}", part_two(&data));
        return;
}

fn part_one(data: &String) -> u64 {
    let re = Regex::new(r"mul\(\d+,\d+\)").unwrap();
    
    let mut ans = 0;
    for mat in re.find_iter(data) {
        let s = mat.as_str();
        let b = s.as_bytes();
        let mut i = 4;
        let mut num1: u64 = 0;
        let mut num2: u64 = 0;
        
        while b[i] as char != ',' {
            num1 = 10 * num1 + ((b[i] - ('0' as u8)) as u64);
            i += 1;
        }
        i += 1;
        while b[i] as char != ')' {
            num2 = 10 * num2 + ((b[i] - ('0' as u8)) as u64);
            i += 1;
        }

        ans += num1 * num2
    }

    ans
}

fn part_two(data: &String) -> u64 {
    let re = Regex::new(r"mul\(\d+,\d+\)|do\(\)|don't\(\)").unwrap();
    
    let mut ans = 0;
    let mut enabled = true;

    for mat in re.find_iter(data) {
        let s = mat.as_str();

        if s == "do()" {
            enabled = true;
            continue;
        }

        if s == "don't()" {
            enabled = false;
            continue;
        }

        if !enabled {
            continue;
        }

        let b = s.as_bytes();
        let mut i = 4;
        let mut num1: u64 = 0;
        let mut num2: u64 = 0;
        
        while b[i] as char != ',' {
            num1 = 10 * num1 + ((b[i] - ('0' as u8)) as u64);
            i += 1;
        }
        i += 1;
        while b[i] as char != ')' {
            num2 = 10 * num2 + ((b[i] - ('0' as u8)) as u64);
            i += 1;
        }

        ans += num1 * num2
    }

    ans
}